kafka/source/pkg/adapter/message.go

```diff
@@ -161,3 +176,79 @@ func getKeyTypeMapper(keyType string) func([]byte) interface{} {
 	}
 	return keyTypeMapper
 }
+
+func getAvroDecoder(schemaURL string) *AvroDecoder {
+	if schemaURL != "" {
+		return &AvroDecoder{
+			schemaURL: schemaURL,
+			codecMap:  make(map[int]*goavro.Codec),
+		}
+	}
+	return nil
+}
+
+func (decoder *AvroDecoder) getCodec(schemaID int) (*goavro.Codec, error) {
+	decoder.mutex.RLock()
+	if decoder.codecMap[schemaID] != nil {
+		return decoder.codecMap[schemaID], nil
+	}
+	decoder.mutex.RUnlock()
+
+	resp, err := nethttp.Get(fmt.Sprintf("%s/schemas/ids/%d", decoder.schemaURL, schemaID))
+	if err != nil {
+		return nil, err
+	}
+
+	if resp.Body != nil {
+		defer resp.Body.Close()
+	}
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	var schemaResp map[string]interface{}
+	if err := json.Unmarshal(body, &schemaResp); err != nil {
+		return nil, err
+	}
+
+	schema, err := json.Marshal(schemaResp["schema"])
+	if err != nil {
+		return nil, err
+	}
+
+	codec, err := goavro.NewCodec(string(schema))
+	if err != nil {
+		return nil, err
+	}
+
+	decoder.mutex.Lock()
+	decoder.codecMap[schemaID] = codec
+	decoder.mutex.Unlock()
+	return codec, nil
+}
+
+func (decoder *AvroDecoder) decode(msg *protocolkafka.Message, logger *zap.Logger) {
```

Comment:
So the point of this decoding using avro is to convert a byte representation of avro to a textual representation (in practice JSON), correct?

Reply:
yes